File: tests/RopExamplesTest.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports NUnit.Framework
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq
Imports tech.janky.dotrop.examples_vb


Namespace tech.janky.dotrop.tests

<TestFixture>
Public Class RopExamplesTest_VB
    Private test_key_ids As List(Of String)

    <OneTimeSetUp>
    Public Sub setUp()
        For Each fname As String In New String() {"pubring.pgp", "secring.pgp"}
            Try
                File.Delete(fname)
            Catch ex As IOException
            End Try
        Next
        test_key_ids = New List(Of String)()
    End Sub
    
    <OneTimeTearDown>
    Public Sub tearDown()
        Dim fnames As List(Of String) = New List(Of String)()
        For Each name As String In New String() {"pubring.pgp", "secring.pgp", "encrypted.asc", "signed.asc"}
            fnames.Add(name)
        Next
        For Each keyid As String In test_key_ids
            fnames.Add(String.Format("key-{0}-pub.asc", keyid))
            fnames.Add(String.Format("key-{0}-sec.asc", keyid))
        Next
        For Each fname As String In fnames
            Try
                File.Delete(fname)
            Catch ex As IOException
            End Try
        Next
    End Sub

    <Test>
    Public Sub test_examples()
        'Execute
        Call New Generate().execute()
        Call New Encrypt().execute()
        Call New Decrypt().execute()
        If Encrypt.message.CompareTo(Decrypt.message) <> 0 Then
            Throw New Exception("Decryption Failed!")
        End If
        Call New Sign().execute()
        For idx As Integer = 0 To 2-1
            test_key_ids.Add(Sign.key_ids(idx))
        Next
        Call New Verify().execute()
        Dim out_() As String = New string() {Nothing}
        Call New Dump().execute(New String() {"Dump", "-j", "signed.asc"}, out_)

        'Parse the dump
        Dim jso As JArray = Nothing, ref_jso As JArray = Nothing
        Try
            jso = JArray.Parse(out_(0))
        Catch ex As JsonReaderException
            Assert.True(False)
        End Try
        
        Dim data As String = Nothing
        Try
        	data = File.ReadAllText("et_json.txt")
        Catch ex As IOException
	        Assert.IsTrue(False)
        End Try
        data = data.Replace("b2617b172b2ceae2a1ed72435fc1286cf91da4d0", Sign.key_fprints(0).ToLower())
        data = data.Replace("5fc1286cf91da4d0", Sign.key_ids(0).ToLower())
        data = data.Replace("f1768c67ec5a9ead3061c2befeee14c57b1a12d9", Sign.key_fprints(1).ToLower())
        data = data.Replace("feee14c57b1a12d9", Sign.key_ids(1).ToLower())
        Try
            ref_jso = JArray.Parse(data)
        Catch ex As JsonReaderException
            Assert.IsTrue(False)
        End Try

        ' Compare the jsons
        right_cmp_json(jso, ref_jso)

        TestContext.Progress.WriteLine("SUCCESS !")
    End Sub
    
    Private Sub right_cmp_json(json As Object, ref_json As Object)
        If TypeOf ref_json Is JArray Then
            For idx As Integer = 0 To ref_json.Count-1
                right_cmp_json(json(idx), ref_json(idx))
            Next
        Else If TypeOf ref_json Is JObject Then
            For Each prop As JProperty In ref_json
                right_cmp_json(json(prop.Name), prop.Value)
            Next
        Else If Not json.Equals(ref_json) Then
            Throw New Exception(String.Format("FAILED! ({0} != {1})", json, ref_json))
        End If
    End Sub
End Class

End Namespace

File: tests/RopExamplesTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using NUnit.Framework;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using tech.janky.dotrop.examples;


namespace tech.janky.dotrop.tests {

[TestFixture]
public class RopExamplesTest {
    private List<string> test_key_ids;

    [OneTimeSetUp]
    public void setUp() {
        foreach(string fname in new String[] {"pubring.pgp", "secring.pgp"}) {
            try {
                File.Delete(fname);
            } catch(IOException) {}
        }
        test_key_ids = new List<string>();
    }
    
    [OneTimeTearDown]
    public void tearDown() {
        List<string> fnames = new List<string>();
        foreach(string name in new string[] {"pubring.pgp", "secring.pgp", "encrypted.asc", "signed.asc"})
            fnames.Add(name);
        foreach(string keyid in test_key_ids) {
            fnames.Add(string.Format("key-{0}-pub.asc", keyid));
            fnames.Add(string.Format("key-{0}-sec.asc", keyid));
        }
        foreach(string fname in fnames) {
            try {
                File.Delete(fname);
            } catch(IOException) {}
        }
    }

    [Test]
    public void test_examples() {
        //Execute
        (new Generate()).execute();
        (new Encrypt()).execute();
        (new Decrypt()).execute();
        if(Encrypt.message.CompareTo(Decrypt.message) != 0)
            throw new Exception("Decryption Failed!");
        (new Sign()).execute();
        for(int idx = 0; idx < 2; idx++)
            test_key_ids.Add(Sign.key_ids[idx]);
        (new Verify()).execute();
        string[] out_ = new string[] {null}; 
        (new Dump()).execute(new string[] {"Dump", "-j", "signed.asc"}, out_);

        //Parse the dump
        JArray jso = null, ref_jso = null;
        try {
            jso = JArray.Parse(out_[0]);
        } catch(JsonReaderException) {
            Assert.True(false);
        }
        
        String data = null;
        try {
        	data = File.ReadAllText("et_json.txt");
        } catch(IOException) {
	        Assert.IsTrue(false);
        }
        data = data.Replace("b2617b172b2ceae2a1ed72435fc1286cf91da4d0", Sign.key_fprints[0].ToLower());
        data = data.Replace("5fc1286cf91da4d0", Sign.key_ids[0].ToLower());
        data = data.Replace("f1768c67ec5a9ead3061c2befeee14c57b1a12d9", Sign.key_fprints[1].ToLower());
        data = data.Replace("feee14c57b1a12d9", Sign.key_ids[1].ToLower());
        try {
            ref_jso = JArray.Parse(data);
        } catch(JsonReaderException) {
            Assert.IsTrue(false);
        }

        // Compare the jsons
        right_cmp_json(jso, ref_jso);

        TestContext.Progress.WriteLine("SUCCESS !");
    }
    
    private void right_cmp_json(object json, object ref_json) {
        if(typeof(JArray).IsInstanceOfType(ref_json))
            for(int idx = 0; idx < ((JArray)ref_json).Count; idx++) 
                right_cmp_json(((JArray)json)[idx], ((JArray)ref_json)[idx]);
        else if(typeof(JObject).IsInstanceOfType(ref_json)) {
            foreach(JProperty prop in ((JObject)ref_json).Properties())
                right_cmp_json(((JObject)json)[prop.Name], prop.Value);
        } else if(!json.Equals(ref_json))
            throw new Exception(string.Format("FAILED! ({0} != {1})", json, ref_json));
    }
}

}
